test_plans/cuda-backend-interop.asciidoc

```diff
@@ -0,0 +1,251 @@
+* `std::size_t`
+* `vec<T, int dim>` where `T` is each of the scalar types listed above except
+   for `bool` and `dim` is `1`, `2`, `3`, `4`, `8`, and `16`.
+* `marray<T, size_t dim>` where `T` is each of the scalar types listed above
+  and `dim` is `2`, `5`, and `10`.
+* A user-defined struct with several scalar member variables, no constructor,
+  destructor or member functions.
+* A user-defined class with several scalar member variables and a user-defined
+  default constructor.
+* A user-defined class with several scalar member variables, a deleted default
+  constructor, and a user-defined (non-default) constructor.
+
+When `T` is the data type for `unsampled_image_accessor` or
+`sampled_image_accessor` it is tested for the following types:
+
+* `sycl::int4`
+* `sycl::uint4`
+* `sycl::float4`
+* `sycl::half4`
+
+== Tests
+
+=== Backend availability
+
+* Check that the macro `SYCL_BACKEND_CUDA` is defined.
+* Check that the `sycl::backend` enum has the member `cuda`.
+* Check that the `sycl::is_backend_active` type specialization for
+`sycl::backend::cuda` is defined with a boolan member variabled called `value`
+which resolves to `true`.
+
+=== Application interoperability
+
+All of the tests in this section are run within SYCL application scope.
+
+==== Check availability of backend interop traits
+
+* Check that the `sycl::backend_input_t<sycl::backend::cuda, sycl::platform>`
+is available and resolves to an alias of `std::vector<CUdevice>`.
+* Check that the `sycl::backend_input_t<sycl::backend::cuda, sycl::device>`
+is available and resolves to an alias of `CUdevice`.
+* Check that the `sycl::backend_input_t<sycl::backend::cuda, sycl::context>`
+is available and resolves to an alias of `std::vector<CUcontext>`.
+* Check that the `sycl::backend_input_t<sycl::backend::cuda, sycl::queue>`
+is available and resolves to an alias of `CUstream`.
+* Check that the `sycl::backend_input_t<sycl::backend::cuda, sycl::event>`
+is available and resolves to an alias of `CUevent`.
+* Check that the `sycl::backend_input_t<sycl::backend::cuda, sycl::buffer>`
+is available and resolves to an alias of
+`struct { CUdeviceptr ptr; size_t size; }`.
+* Check that the `sycl::backend_return_t<sycl::backend::cuda, sycl::platform>`
+is available and resolves to an alias of `std::vector<CUdevice>`.
+* Check that the `sycl::backend_return_t<sycl::backend::cuda, sycl::device>`
+is available and resolves to an alias of `CUdevice`.
+* Check that the `sycl::backend_return_t<sycl::backend::cuda, sycl::context>`
+is available and resolves to an alias of `std::vector<CUcontext>`.
+* Check that the `sycl::backend_return_t<sycl::backend::cuda, sycl::queue>`
+is available and resolves to an alias of `CUstream`.
+* Check that the `sycl::backend_return_t<sycl::backend::cuda, sycl::event>`
+is available and resolves to an alias of `CUevent`.
+* Check that the `sycl::backend_return_t<sycl::backend::cuda, sycl::buffer>`
+is available and resolves to an alias of `CUdeviceptr`.
```

Comment:
At least it is interesting to see this as a *Rosetta Stone* for SYCL-CUDA translation. :-)